### PR TITLE
Implement hex-to-base64 encoding migration for KeyPackage and Welcome

### PR DESCRIPTION
**Pull request by mubarakcoded**
This implements a 3-phase migration strategy to reduce payload sizes by ~33%:

Phase 1 (Current):
- Add dual-format reading support (both hex and base64)
- Keep writing hex by default (use_base64_encoding: false)
- Backward and forward compatible

Phase 2 (After 3 months):
- Switch to base64 encoding via config flag
- Still read both formats during transition

Phase 3 (After 6 months - Optional):
- Remove hex support entirely

Changes:
- Add MdkConfig struct with use_base64_encoding flag (default: false)
- Implement smart format detection based on character set
  - Hex: only [0-9a-fA-F]
  - Base64: [A-Za-z0-9+/=]
- Add conditional encoding in key package and welcome creation
- Update builder pattern to support configuration

This ensures a smooth migration that handles the 3-month key package lifetime without breaking compatibility.

## Summary by CodeRabbit

* **New Features**
  * Builder-style MDK configuration with a toggle for base64 vs hex encoding; runtime-selectable encoding for outgoing payloads and encoding-aware decoding for incoming welcomes/key packages (encoding indicated via tags).

* **Bug Fixes**
  * Clearer, encoding-aware decoding errors and persisted failure state when decoding fails.

* **Public API**
  * Updated public returns and builder/config exposure to support encoding-tagged payloads and variable-length tags.

* **Tests**
  * Expanded tests for cross-format decoding, invalid inputs, padding/edge cases, and size differences.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Review by dannym-arx — APPROVED** [review]
Looks great to me!

My only comment is whether it's even possible to correctly detect if something is hex or base64 with 100% accuracy

c0ffeebabe is valid hex, but also valid base64 of course.

perhaps a version tag makes more sense (no tag = hex, v1 = base64)? but I'll leave this as an open question

File: crates/mdk-core/src/key_packages.rs [hunk 1]
```diff
@@ -11,28 +11,33 @@ use tls_codec::{Deserialize as TlsDeserialize, Serialize as TlsSerialize};
...
 where
     Storage: MdkStorageProvider,
 {
     /// Creates a key package for a Nostr event.
     ///
-    /// This function generates a hex-encoded key package that is used as the content field of a kind:443 Nostr event.
+    /// This function generates an encoded key package that is used as the content field of a kind:443 Nostr event.
+    /// The encoding format is determined by `MdkConfig::use_base64_encoding`:
+    /// - When `false` (default): uses hex encoding (legacy format, no encoding tag)
+    /// - When `true`: uses base64 encoding with `["encoding", "base64"]` tag (~33% smaller)
```
[the remark this answers is not in the document]
Comment (dannym-arx) [reply crates/mdk-core/src/key_packages.rs]: $$P_{\text{ambiguous}}(\text{msglen}) = \left(\frac{22}{64}\right)^{\text{msglen}} \approx 0.344^\text{msglen}$$

File: crates/mdk-core/src/key_packages.rs [hunk 2]
```diff
@@ -2029,4 +2109,333 @@ mod tests {
...
+        let test_pubkey =
+            PublicKey::from_hex("884704bd421671e01c13f854d2ce23ce2a5bfe9562f4f297ad2bc921ba30c3a6")
+                .unwrap();
+        let relays = vec![RelayUrl::parse("wss://relay.example.com").unwrap()];
+
+        let (key_package_str, tags) = mdk
+            .create_key_package_for_event(&test_pubkey, relays)
+            .expect("Failed to create key package");
+
+        // Verify it's valid hex
```
Comment (jgmontoya) [inline crates/mdk-core/src/key_packages.rs]: same here

File: crates/mdk-core/src/key_packages.rs [hunk 3]
```diff
@@ -2029,4 +2109,333 @@ mod tests {
...
+        assert!(hex::decode(&key_package_str).is_ok(), "Should be valid hex");
+
+        // Verify no encoding tag is present (hex is the default)
+        let encoding_tag = tags
+            .iter()
+            .find(|t| t.as_slice().first() == Some(&"encoding".to_string()));
+        assert!(
+            encoding_tag.is_none(),
+            "Should NOT have encoding tag when using hex (default)"
+        );
```
Comment (jgmontoya) [inline crates/mdk-core/src/key_packages.rs]: Related to the prior comments, I think we should still include the tag (and verify that we're doing so here in the tests)

File: crates/mdk-core/src/key_packages.rs [hunk 4]
```diff
@@ -2029,4 +2109,333 @@ mod tests {
...
+
+        // Base64 should be approximately 33% smaller (hex is 2x, base64 is 1.33x)
+        // Allow some variance due to encoding overhead
+        assert!(
+            savings_percent > 25.0 && savings_percent < 40.0,
+            "Expected savings between 25-40%, got {:.1}%",
+            savings_percent
+        );
+    }
```
Comment (jgmontoya) [inline crates/mdk-core/src/key_packages.rs]: I don't think this test is really useful. This is basically testing a property of the encodings themselves and not our code. I'd just remove it entirely

File: crates/mdk-core/src/key_packages.rs [hunk 5]
```diff
@@ -2029,4 +2109,333 @@ mod tests {
...
+    #[test]
+    fn test_decode_valid_hex() {
+        let mdk = create_test_mdk();
+
+        // Test that valid hex strings decode successfully
+        let valid_hex = "00000000";
+        let result = mdk.decode_key_package_content(valid_hex, ContentEncoding::Hex);
+        assert!(result.is_ok(), "Valid hex should decode successfully");
+        assert_eq!(result.unwrap(), vec![0, 0, 0, 0]);
+    }
```
Comment (jgmontoya) [inline crates/mdk-core/src/key_packages.rs]: This test seems a bit redundant given we have the test_key_package_hex_encoding_legacy test

**Review by jgmontoya — COMMENTED** [review]
Some comments, let's try to perhaps address them on a future PR
